{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Final Project Katherine Maki - BIOF 309"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Project Question: How many bacterial features (at the genus level) are unique and overlapping across three major taxonomy databases?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What is a bacterial genus?\n",
    "\n",
    "Bacteria is classified taxonomically ranging from Phylum (high level) to species strain"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![taxonomy](img/Taxonomy_copy2.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In microbiome analysis, 16S rRNA gene amplicon sequencing is a \"culture independent\" analysis technique used to classify bacteria based on hypervariable regions in the bacterial 16S rRNA gene. Bacteria is classified based on genetic sequences that are amplified from DNA extracted from a sample. For example, to noninvasively analyze the gut microbiome you can collect and extract DNA from fecal samples :)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Four Databases that will be analyzed\n",
    "#### Ribosomal Database Project\n",
    "#### Greengenes\n",
    "#### Silva \n",
    "#### Human Oral Microbiome Database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Links to data\n",
    "[Human Oral Microbiome Database](http://www.homd.org/?name=Download&taxonly=1)\n",
    "[Silva Release 138](https://www.arb-silva.de/no_cache/download/archive/release_138/Exports/)\n",
    "[Greengenes 13_8](https://docs.qiime2.org/2020.2/data-resources/)\n",
    "[Ribosomal Database Project V16](https://mothur.org/wiki/rdp_reference_files/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Since there are four datasets and they are all formatted differently, we will bring them in seperately and format them to extra characters are deleted, empty columns are changed to \"NaN\" and the genus data is subsetted"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Greengenes is imported first"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import GreenGenes Dataset without header\n",
    "gg = pd.read_table('data_input/gg_99_otu_taxonomy.txt', header=None, names=['id', 'taxonomy'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Split Single Taxonomy Column into Several Columns\n",
    "gg = gg.taxonomy.str.split(\";\",expand=True)\n",
    "gg.columns = ['Kingdom','Phylum','Class',\n",
    "                     'Order','Family','Genus', 'Species']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get rid of the preceding type labels \n",
    "gg.Kingdom = gg.Kingdom.apply(lambda x: x.replace('k__',''))\n",
    "gg.Phylum = gg.Phylum.apply(lambda x: x.replace('p__',''))\n",
    "gg.Class = gg.Class.apply(lambda x: x.replace('c__',''))\n",
    "gg.Order = gg.Order.apply(lambda x: x.replace('o__',''))\n",
    "gg.Family = gg.Family.apply(lambda x: x.replace('f__',''))\n",
    "gg.Genus = gg.Genus.apply(lambda x: x.replace('g__',''))\n",
    "gg.Species = gg.Species.apply(lambda x: x.replace('s__',''))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Replace blank spaces with NaN\n",
    "gg = gg.replace(r'^\\s*$', np.nan, regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add column identifying the dataset\n",
    "gg['Dataset'] = 'Greengenes'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>Dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Synechococcophycideae</td>\n",
       "      <td>Synechococcales</td>\n",
       "      <td>Synechococcaceae</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Greengenes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>Pelagibacteraceae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Greengenes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Greengenes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Bacillales</td>\n",
       "      <td>Staphylococcaceae</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Greengenes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Bacillales</td>\n",
       "      <td>Bacillaceae</td>\n",
       "      <td>Anoxybacillus</td>\n",
       "      <td>kestanbolensis</td>\n",
       "      <td>Greengenes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Kingdom           Phylum                   Class             Order  \\\n",
       "0  Bacteria    Cyanobacteria   Synechococcophycideae   Synechococcales   \n",
       "1  Bacteria   Proteobacteria     Alphaproteobacteria     Rickettsiales   \n",
       "2  Bacteria   Actinobacteria          Actinobacteria   Actinomycetales   \n",
       "3  Bacteria       Firmicutes                 Bacilli        Bacillales   \n",
       "4  Bacteria       Firmicutes                 Bacilli        Bacillales   \n",
       "\n",
       "               Family            Genus          Species     Dataset  \n",
       "0    Synechococcaceae    Synechococcus              NaN  Greengenes  \n",
       "1   Pelagibacteraceae              NaN              NaN  Greengenes  \n",
       "2    Mycobacteriaceae    Mycobacterium              NaN  Greengenes  \n",
       "3   Staphylococcaceae   Staphylococcus              NaN  Greengenes  \n",
       "4         Bacillaceae    Anoxybacillus   kestanbolensis  Greengenes  "
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#View output\n",
    "gg.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Pull out Genus and Dataset Variables\n",
    "gg_genus =  gg[['Genus', 'Dataset']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genus</th>\n",
       "      <th>Dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>Greengenes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Greengenes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Greengenes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>Greengenes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Anoxybacillus</td>\n",
       "      <td>Greengenes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Genus     Dataset\n",
       "0    Synechococcus  Greengenes\n",
       "1              NaN  Greengenes\n",
       "2    Mycobacterium  Greengenes\n",
       "3   Staphylococcus  Greengenes\n",
       "4    Anoxybacillus  Greengenes"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gg_genus.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Next Silva is imported"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import GreenGenes Dataset without header\n",
    "silva = pd.read_table('data_input/silva_consensus_taxonomy_7_levels.txt', header=None, names=['id', 'taxonomy'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "silva = silva.taxonomy.str.split(\";\",expand=True)\n",
    "silva.columns = ['Kingdom','Phylum','Class',\n",
    "                     'Order','Family','Genus', 'Species']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get rid of the preceding type labels \n",
    "silva.Kingdom = silva.Kingdom.apply(lambda x: x.replace('D_0__',''))\n",
    "silva.Phylum = silva.Phylum.apply(lambda x: x.replace('D_1__',''))\n",
    "silva.Class = silva.Class.apply(lambda x: x.replace('D_2__',''))\n",
    "silva.Order = silva.Order.apply(lambda x: x.replace('D_3__',''))\n",
    "silva.Family = silva.Family.apply(lambda x: x.replace('D_4__',''))\n",
    "silva.Genus = silva.Genus.apply(lambda x: x.replace('D_5__',''))\n",
    "silva.Species = silva.Species.apply(lambda x: x.replace('D_6__',''))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Epsilonbacteraeota</td>\n",
       "      <td>Campylobacteria</td>\n",
       "      <td>Campylobacterales</td>\n",
       "      <td>Thiovulaceae</td>\n",
       "      <td>Sulfuricurvum</td>\n",
       "      <td>Sulfuricurvum sp. EW1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Oxyphotobacteria</td>\n",
       "      <td>Nostocales</td>\n",
       "      <td>Nostocaceae</td>\n",
       "      <td>Nostoc PCC-73102</td>\n",
       "      <td>Nostoc sp. 'Nephroma expallidum cyanobiont 23'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Bacteroidales</td>\n",
       "      <td>Muribaculaceae</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Alteromonadales</td>\n",
       "      <td>Pseudoalteromonadaceae</td>\n",
       "      <td>Pseudoalteromonas</td>\n",
       "      <td>Pseudoalteromonas sp.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacteriales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>uncultured organism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Clostridiales</td>\n",
       "      <td>Ruminococcaceae</td>\n",
       "      <td>Ruminococcaceae UCG-005</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Clostridiales</td>\n",
       "      <td>Ruminococcaceae</td>\n",
       "      <td>Ruminococcaceae V9D2013 group</td>\n",
       "      <td>uncultured rumen bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Thermolithobacterales</td>\n",
       "      <td>Thermolithobacteraceae</td>\n",
       "      <td>Thermolithobacter</td>\n",
       "      <td>Thermolithobacter ferrireducens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Streptomycetales</td>\n",
       "      <td>Streptomycetaceae</td>\n",
       "      <td>Streptomyces</td>\n",
       "      <td>Streptomyces luteogriseus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Bacteroidales</td>\n",
       "      <td>Porphyromonadaceae</td>\n",
       "      <td>Porphyromonas</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Tenericutes</td>\n",
       "      <td>Mollicutes</td>\n",
       "      <td>Entomoplasmatales</td>\n",
       "      <td>Spiroplasmataceae</td>\n",
       "      <td>Spiroplasma</td>\n",
       "      <td>Spiroplasma leptinotarsae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Flavobacteriales</td>\n",
       "      <td>Crocinitomicaceae</td>\n",
       "      <td>Fluviicola</td>\n",
       "      <td>Ambiguous_taxa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Deltaproteobacteria</td>\n",
       "      <td>Bdellovibrionales</td>\n",
       "      <td>Bdellovibrionaceae</td>\n",
       "      <td>Bdellovibrio</td>\n",
       "      <td>uncultured Bdellovibrio sp.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Clostridiales</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Tyzzerella 3</td>\n",
       "      <td>uncultured Firmicutes bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Crenarchaeota</td>\n",
       "      <td>Bathyarchaeia</td>\n",
       "      <td>uncultured archaeon</td>\n",
       "      <td>uncultured archaeon</td>\n",
       "      <td>uncultured archaeon</td>\n",
       "      <td>uncultured archaeon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Micromonosporales</td>\n",
       "      <td>Micromonosporaceae</td>\n",
       "      <td>Micromonospora</td>\n",
       "      <td>Micromonospora sp. 162022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Rokubacteria</td>\n",
       "      <td>NC10</td>\n",
       "      <td>Methylomirabilales</td>\n",
       "      <td>Methylomirabilaceae</td>\n",
       "      <td>wb1-A12</td>\n",
       "      <td>Ambiguous_taxa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Clostridiales</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Lachnospiraceae AC2044 group</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Carnobacteriaceae</td>\n",
       "      <td>Carnobacterium</td>\n",
       "      <td>Carnobacterium sp. Y6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacteriales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Budvicia</td>\n",
       "      <td>Budvicia diplopodorum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Caulobacterales</td>\n",
       "      <td>Caulobacteraceae</td>\n",
       "      <td>Brevundimonas</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>21</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiomicrospirales</td>\n",
       "      <td>Thiomicrospiraceae</td>\n",
       "      <td>Thiomicrorhabdus</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>22</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Bacteroidales</td>\n",
       "      <td>M2PB4-65 termite group</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>23</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Chromatiales</td>\n",
       "      <td>Chromatiaceae</td>\n",
       "      <td>Chromatium</td>\n",
       "      <td>uncultured Chromatiaceae bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Negativicutes</td>\n",
       "      <td>Selenomonadales</td>\n",
       "      <td>Veillonellaceae</td>\n",
       "      <td>uncultured</td>\n",
       "      <td>uncultured Firmicutes bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>25</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Oxyphotobacteria</td>\n",
       "      <td>Chloroplast</td>\n",
       "      <td>Lambia antarctica</td>\n",
       "      <td>Lambia antarctica</td>\n",
       "      <td>Lambia antarctica</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>26</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Parvibaculales</td>\n",
       "      <td>PS1 clade</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>27</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Oxyphotobacteria</td>\n",
       "      <td>Chloroplast</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>28</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Rhodanobacteraceae</td>\n",
       "      <td>Dyella</td>\n",
       "      <td>Dyella sp. ECB15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>29</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>Thermoplasmatales</td>\n",
       "      <td>Aciduliprofundaceae</td>\n",
       "      <td>Aciduliprofundum</td>\n",
       "      <td>Ambiguous_taxa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>30</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacteriales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Serratia</td>\n",
       "      <td>uncultured actinomycete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>31</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Chloroflexi</td>\n",
       "      <td>Anaerolineae</td>\n",
       "      <td>Ardenticatenales</td>\n",
       "      <td>uncultured</td>\n",
       "      <td>uncultured organism</td>\n",
       "      <td>uncultured organism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>32</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Flavobacteriales</td>\n",
       "      <td>Schleiferiaceae</td>\n",
       "      <td>Schleiferia</td>\n",
       "      <td>uncultured organism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>33</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Bacteroidales</td>\n",
       "      <td>Prolixibacteraceae</td>\n",
       "      <td>WCHB1-32</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>34</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Bacteroidales</td>\n",
       "      <td>Rikenellaceae</td>\n",
       "      <td>Alistipes</td>\n",
       "      <td>uncultured Alistipes sp.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>35</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Alteromonadales</td>\n",
       "      <td>Pseudoalteromonadaceae</td>\n",
       "      <td>Pseudoalteromonas</td>\n",
       "      <td>Pseudoalteromonas telluritireducens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>36</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Planctomycetes</td>\n",
       "      <td>Planctomycetacia</td>\n",
       "      <td>Planctomycetales</td>\n",
       "      <td>Rubinisphaeraceae</td>\n",
       "      <td>SH-PL14</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>37</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Asgardaeota</td>\n",
       "      <td>Lokiarchaeia</td>\n",
       "      <td>uncultured archaeon</td>\n",
       "      <td>uncultured archaeon</td>\n",
       "      <td>uncultured archaeon</td>\n",
       "      <td>uncultured archaeon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>38</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Planctomycetes</td>\n",
       "      <td>Planctomycetacia</td>\n",
       "      <td>Pirellulales</td>\n",
       "      <td>Pirellulaceae</td>\n",
       "      <td>Pir4 lineage</td>\n",
       "      <td>metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>39</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Cellvibrionales</td>\n",
       "      <td>Porticoccaceae</td>\n",
       "      <td>SAR92 clade</td>\n",
       "      <td>uncultured marine bacterium</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Kingdom              Phylum                Class                  Order  \\\n",
       "0   Bacteria  Epsilonbacteraeota      Campylobacteria      Campylobacterales   \n",
       "1   Bacteria       Cyanobacteria     Oxyphotobacteria             Nostocales   \n",
       "2   Bacteria       Bacteroidetes          Bacteroidia          Bacteroidales   \n",
       "3   Bacteria      Proteobacteria  Gammaproteobacteria        Alteromonadales   \n",
       "4   Bacteria      Proteobacteria  Gammaproteobacteria      Enterobacteriales   \n",
       "5   Bacteria          Firmicutes           Clostridia          Clostridiales   \n",
       "6   Bacteria          Firmicutes           Clostridia          Clostridiales   \n",
       "7   Bacteria          Firmicutes           Clostridia  Thermolithobacterales   \n",
       "8   Bacteria      Actinobacteria       Actinobacteria       Streptomycetales   \n",
       "9   Bacteria       Bacteroidetes          Bacteroidia          Bacteroidales   \n",
       "10  Bacteria         Tenericutes           Mollicutes      Entomoplasmatales   \n",
       "11  Bacteria       Bacteroidetes          Bacteroidia       Flavobacteriales   \n",
       "12  Bacteria      Proteobacteria  Deltaproteobacteria      Bdellovibrionales   \n",
       "13  Bacteria          Firmicutes           Clostridia          Clostridiales   \n",
       "14   Archaea       Crenarchaeota        Bathyarchaeia    uncultured archaeon   \n",
       "15  Bacteria      Actinobacteria       Actinobacteria      Micromonosporales   \n",
       "16  Bacteria        Rokubacteria                 NC10     Methylomirabilales   \n",
       "17  Bacteria          Firmicutes           Clostridia          Clostridiales   \n",
       "18  Bacteria          Firmicutes              Bacilli        Lactobacillales   \n",
       "19  Bacteria      Proteobacteria  Gammaproteobacteria      Enterobacteriales   \n",
       "20  Bacteria      Proteobacteria  Alphaproteobacteria        Caulobacterales   \n",
       "21  Bacteria      Proteobacteria  Gammaproteobacteria      Thiomicrospirales   \n",
       "22  Bacteria       Bacteroidetes          Bacteroidia          Bacteroidales   \n",
       "23  Bacteria      Proteobacteria  Gammaproteobacteria           Chromatiales   \n",
       "24  Bacteria          Firmicutes        Negativicutes        Selenomonadales   \n",
       "25  Bacteria       Cyanobacteria     Oxyphotobacteria            Chloroplast   \n",
       "26  Bacteria      Proteobacteria  Alphaproteobacteria         Parvibaculales   \n",
       "27  Bacteria       Cyanobacteria     Oxyphotobacteria            Chloroplast   \n",
       "28  Bacteria      Proteobacteria  Gammaproteobacteria        Xanthomonadales   \n",
       "29   Archaea       Euryarchaeota       Thermoplasmata      Thermoplasmatales   \n",
       "30  Bacteria      Proteobacteria  Gammaproteobacteria      Enterobacteriales   \n",
       "31  Bacteria         Chloroflexi         Anaerolineae       Ardenticatenales   \n",
       "32  Bacteria       Bacteroidetes          Bacteroidia       Flavobacteriales   \n",
       "33  Bacteria       Bacteroidetes          Bacteroidia          Bacteroidales   \n",
       "34  Bacteria       Bacteroidetes          Bacteroidia          Bacteroidales   \n",
       "35  Bacteria      Proteobacteria  Gammaproteobacteria        Alteromonadales   \n",
       "36  Bacteria      Planctomycetes     Planctomycetacia       Planctomycetales   \n",
       "37   Archaea         Asgardaeota         Lokiarchaeia    uncultured archaeon   \n",
       "38  Bacteria      Planctomycetes     Planctomycetacia           Pirellulales   \n",
       "39  Bacteria      Proteobacteria  Gammaproteobacteria        Cellvibrionales   \n",
       "\n",
       "                    Family                          Genus  \\\n",
       "0             Thiovulaceae                  Sulfuricurvum   \n",
       "1              Nostocaceae               Nostoc PCC-73102   \n",
       "2           Muribaculaceae           uncultured bacterium   \n",
       "3   Pseudoalteromonadaceae              Pseudoalteromonas   \n",
       "4       Enterobacteriaceae                     Klebsiella   \n",
       "5          Ruminococcaceae        Ruminococcaceae UCG-005   \n",
       "6          Ruminococcaceae  Ruminococcaceae V9D2013 group   \n",
       "7   Thermolithobacteraceae              Thermolithobacter   \n",
       "8        Streptomycetaceae                   Streptomyces   \n",
       "9       Porphyromonadaceae                  Porphyromonas   \n",
       "10       Spiroplasmataceae                    Spiroplasma   \n",
       "11       Crocinitomicaceae                     Fluviicola   \n",
       "12      Bdellovibrionaceae                   Bdellovibrio   \n",
       "13         Lachnospiraceae                   Tyzzerella 3   \n",
       "14     uncultured archaeon            uncultured archaeon   \n",
       "15      Micromonosporaceae                 Micromonospora   \n",
       "16     Methylomirabilaceae                        wb1-A12   \n",
       "17         Lachnospiraceae   Lachnospiraceae AC2044 group   \n",
       "18       Carnobacteriaceae                 Carnobacterium   \n",
       "19      Enterobacteriaceae                       Budvicia   \n",
       "20        Caulobacteraceae                  Brevundimonas   \n",
       "21      Thiomicrospiraceae               Thiomicrorhabdus   \n",
       "22  M2PB4-65 termite group           uncultured bacterium   \n",
       "23           Chromatiaceae                     Chromatium   \n",
       "24         Veillonellaceae                     uncultured   \n",
       "25       Lambia antarctica              Lambia antarctica   \n",
       "26               PS1 clade           uncultured bacterium   \n",
       "27    uncultured bacterium           uncultured bacterium   \n",
       "28      Rhodanobacteraceae                         Dyella   \n",
       "29     Aciduliprofundaceae               Aciduliprofundum   \n",
       "30      Enterobacteriaceae                       Serratia   \n",
       "31              uncultured            uncultured organism   \n",
       "32         Schleiferiaceae                    Schleiferia   \n",
       "33      Prolixibacteraceae                       WCHB1-32   \n",
       "34           Rikenellaceae                      Alistipes   \n",
       "35  Pseudoalteromonadaceae              Pseudoalteromonas   \n",
       "36       Rubinisphaeraceae                        SH-PL14   \n",
       "37     uncultured archaeon            uncultured archaeon   \n",
       "38           Pirellulaceae                   Pir4 lineage   \n",
       "39          Porticoccaceae                    SAR92 clade   \n",
       "\n",
       "                                           Species  \n",
       "0                            Sulfuricurvum sp. EW1  \n",
       "1   Nostoc sp. 'Nephroma expallidum cyanobiont 23'  \n",
       "2                             uncultured bacterium  \n",
       "3                            Pseudoalteromonas sp.  \n",
       "4                              uncultured organism  \n",
       "5                             uncultured bacterium  \n",
       "6                       uncultured rumen bacterium  \n",
       "7                  Thermolithobacter ferrireducens  \n",
       "8                        Streptomyces luteogriseus  \n",
       "9                             uncultured bacterium  \n",
       "10                       Spiroplasma leptinotarsae  \n",
       "11                                  Ambiguous_taxa  \n",
       "12                     uncultured Bdellovibrio sp.  \n",
       "13                 uncultured Firmicutes bacterium  \n",
       "14                             uncultured archaeon  \n",
       "15                       Micromonospora sp. 162022  \n",
       "16                                  Ambiguous_taxa  \n",
       "17                            uncultured bacterium  \n",
       "18                           Carnobacterium sp. Y6  \n",
       "19                           Budvicia diplopodorum  \n",
       "20                            uncultured bacterium  \n",
       "21                            uncultured bacterium  \n",
       "22                            uncultured bacterium  \n",
       "23              uncultured Chromatiaceae bacterium  \n",
       "24                 uncultured Firmicutes bacterium  \n",
       "25                               Lambia antarctica  \n",
       "26                            uncultured bacterium  \n",
       "27                            uncultured bacterium  \n",
       "28                                Dyella sp. ECB15  \n",
       "29                                  Ambiguous_taxa  \n",
       "30                         uncultured actinomycete  \n",
       "31                             uncultured organism  \n",
       "32                             uncultured organism  \n",
       "33                            uncultured bacterium  \n",
       "34                        uncultured Alistipes sp.  \n",
       "35             Pseudoalteromonas telluritireducens  \n",
       "36                            uncultured bacterium  \n",
       "37                             uncultured archaeon  \n",
       "38                                      metagenome  \n",
       "39                     uncultured marine bacterium  "
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "silva.head(40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 369953 entries, 0 to 369952\n",
      "Data columns (total 7 columns):\n",
      "Kingdom    369953 non-null object\n",
      "Phylum     369953 non-null object\n",
      "Class      369953 non-null object\n",
      "Order      369953 non-null object\n",
      "Family     369953 non-null object\n",
      "Genus      369953 non-null object\n",
      "Species    369953 non-null object\n",
      "dtypes: object(7)\n",
      "memory usage: 19.8+ MB\n"
     ]
    }
   ],
   "source": [
    "silva.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genus</th>\n",
       "      <th>Dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>count</td>\n",
       "      <td>93463</td>\n",
       "      <td>203452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>unique</td>\n",
       "      <td>2062</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>top</td>\n",
       "      <td>Bacteroides</td>\n",
       "      <td>Greengenes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>freq</td>\n",
       "      <td>2747</td>\n",
       "      <td>203452</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Genus     Dataset\n",
       "count          93463      203452\n",
       "unique          2062           1\n",
       "top      Bacteroides  Greengenes\n",
       "freq            2747      203452"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gg_genus.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2) Provide a link to your data\n",
    "Your data is required to be free and open to anyone.\n",
    "As such, you should have a URL which anyone can use to download your data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Enter link here."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3) Import your data\n",
    "In the space below, import your data.\n",
    "If your data span multiple files, read them all in.\n",
    "If applicable, merge or append them as needed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4) Show me the head of your data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5) Show me the shape of your data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6) Show me the proportion of missing observations for each column of your data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7) Give me a problem statement.\n",
    "Below, write a problem statement. Keep in mind that your task is to tease out relationships in your data and eventually build a predictive model. Your problem statement can be vague, but you should have a goal in mind. Your problem statement should be between one sentence and one paragraph."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 8) What is your _y_-variable?\n",
    "For final project, you will need to perform a statistical model. This means you will have to accurately predict some y-variable for some combination of x-variables. From your problem statement in part 7, what is that y-variable?"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
